web3/middleware/validation.py

```diff
@@ -27,20 +41,72 @@ def validate_chain_id(web3, chain_id):
         )
 
 
+def check_extradata_length(val):
+    if not isinstance(val, (str, int, bytes)):
+        return val
+    result = HexBytes(val)
+    if len(result) > MAX_EXTRADATA_LENGTH:
+        raise ValidationError(
+            "The field extraData is %d bytes, but should be %d. "
+            "It is quite likely that you are connected to a POA chain. "
+            "Refer "
+            "http://web3py.readthedocs.io/en/latest/middleware.html#geth-style-proof-of-authority "
+            "for more details. The full extraData is: %r" % (
+                len(result), MAX_EXTRADATA_LENGTH, result
+            )
+        )
+    return val
+
+
 def transaction_normalizer(transaction):
     return dissoc(transaction, 'chainId')
 
 
-def validation_middleware(make_request, web3):
-    transaction_validator = apply_formatters_to_dict({
-    })
+def transaction_param_validator(web3):
+    transactions_params_validators = {
+        'chainId': apply_formatter_if(
+            # Bypass `validate_chain_id` if chainId can't be determined
+            lambda _: is_not_null(web3.net.chainId),
+            validate_chain_id(web3)
+        ),
+    }
+    return apply_formatter_at_index(
+        apply_formatters_to_dict(transactions_params_validators),
+        0
+    )
+
+
+BLOCK_VALIDATORS = {
+    'extraData': check_extradata_length,
+}
+
+
+block_validator = apply_formatters_to_dict(BLOCK_VALIDATORS)
+
+
+@curry
+def chain_id_validator(web3):
+    return compose(
+        apply_formatter_at_index(transaction_normalizer, 0),
+        transaction_param_validator(web3)
+    )
+
+
+extra_data_validator = apply_formatter_if(is_not_null, block_validator)
```

Comment:
Since this is intended to be a full block validator, the name `extra_data_validator` feels a bit off. Rather than come up with yet another name, it could be reasonable to combine this with the earlier `block_validator` definition:

```py
block_validator = apply_formatter_if(
    is_not_null,
    apply_formatters_to_dict(BLOCK_VALIDATORS),
)
```